crates/interpreter/src/side_table.rs

```diff
@@ -12,14 +12,59 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use core::slice;
+
+use bytemuck::cast;
+
 use crate::bit_field::*;
 use crate::error::*;
+use crate::parser;
+use crate::toctou::Use;
+
+pub type Parser<'m> = parser::Parser<'m, Use>;
+
+#[allow(dead_code)]
+pub struct SideTable<'m> {
+    indices: &'m [u16], // including 0 and the length of metadata_array
+    metadata: &'m [u16],
+}
+
+#[allow(dead_code)]
+impl<'m> SideTable<'m> {
+    fn metadata(&self, func_idx: usize) -> Metadata<'m> {
+        Metadata(
+            &self.metadata[self.indices[func_idx] as usize .. self.indices[func_idx + 1] as usize],
+        )
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Copy, Clone)]
+struct Metadata<'m>(&'m [u16]);
+
+#[allow(dead_code)]
+impl<'m> Metadata<'m> {
+    pub fn type_idx(&self) -> usize {
+        self.0[0] as usize
+    }
+
+    pub fn parser(&self, code: &'m [u8]) -> Parser<'m> {
+        unsafe { Parser::new(&code[self.0[1] as usize .. self.0[5] as usize]) }
```

Comment:
There's a few issues here:
- We need to read 2 u16 to get a u32.
- We need to read 1..3 and 3..5 (instead of 1 and 5)

We can probably use something like `bytemuck::pod_read_unaligned::<u32>(bytemuck::bytes_of(&self.0[1..3]))` to read a u32 from a &[u16] of length 2.